File: clients/javascript/model/specialty.js
/**
 * Player's specialty
 */
class Specialty {
    constructor(name, tag) {
        this.name = name;
        this.tag = tag;
    }

    /**
     * Logistics. Increased travel distance
     */
    static LOGISTICS = new Specialty("LOGISTICS", 0);
    /**
     * Production. Increased work speed
     */
    static PRODUCTION = new Specialty("PRODUCTION", 1);
    /**
     * Combat. Increased damage
     */
    static COMBAT = new Specialty("COMBAT", 2);

    /**
     * Read Specialty from input stream
     */
    static async readFrom(stream) {
        const tag = await stream.readInt();
        if (tag == Specialty.LOGISTICS.tag) {
            return Specialty.LOGISTICS;
        }
        if (tag == Specialty.PRODUCTION.tag) {
            return Specialty.PRODUCTION;
        }
        if (tag == Specialty.COMBAT.tag) {
            return Specialty.COMBAT;
        }
        throw new Error("Unexpected tag value");
    }

    /**
     * Write Specialty to output stream
     */
    async writeTo(stream) {
        await stream.writeInt(this.tag);
    }

    [Symbol.for('nodejs.util.inspect.custom')]() {
        return this.name;
    }
}

module.exports = Specialty;

File: clients/rust/src/model/specialty.rs
use super::*;

/// Player's specialty
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum Specialty {
    /// Logistics. Increased travel distance
    Logistics,
    /// Production. Increased work speed
    Production,
    /// Combat. Increased damage
    Combat,
}

impl trans::Trans for Specialty {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        let tag: i32 = match self {
            Self::Logistics => 0,
            Self::Production => 1,
            Self::Combat => 2,
        };
        trans::Trans::write_to(&tag, writer)
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let tag = <i32 as trans::Trans>::read_from(reader)?;
        match tag {
            0 => Ok(Self::Logistics),
            1 => Ok(Self::Production),
            2 => Ok(Self::Combat),
            _ => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Unexpected tag {:?}", tag))),
        }
    }
}